cranelift/codegen/src/ir/pcc.rs

```diff
@@ -0,0 +1,347 @@
+//! through optimizations and through lowering. They are thus also
+//! present on VRegs in the VCode. In theory, facts could be checked
+//! at either level, though in practice it is most useful to check
+//! them at the VCode level if the goal is an end-to-end verification
+//! of certain properties (e.g., memory sandboxing).
+//!
+//! Checking facts entails visiting each instruction that defines a
+//! value with a fact, and checking the result's fact against the
+//! facts on arguments and the operand. For VCode, this is
+//! fundamentally a question of the target ISA's semantics, so we call
+//! into the `LowerBackend` for this. Note that during checking there
+//! is also limited forward propagation / inference, but only within
+//! an instruction: for example, an addressing mode commonly can
+//! include an addition, multiplication/shift, or extend operation,
+//! and there is no way to attach facts to the intermediate values
+//! "inside" the instruction, so instead the backend can use
+//! `Fact::add()` and friends to forward-propagate facts.
+//!
+//! TODO:
+//! - Propagate facts through optimization (egraph layer).
+//! - Generate facts in cranelift-wasm frontend when lowering memory ops.
+//! - Implement richer "points-to" facts that describe the pointed-to
+//!   memory, so the loaded values can also have facts.
+//! - Support bounds-checking-type operations for dynamic memories and
+//!   tables.
+//! - Implement checking at the CLIF level as well.
+//! - Check instructions that can trap as well?
+
+use crate::ir;
+use crate::isa::TargetIsa;
+use crate::machinst::{InsnIndex, LowerBackend, MachInst, VCode};
+use std::fmt;
+
+/// The result of checking proof-carrying-code facts.
+pub type PccResult<T> = std::result::Result<T, PccError>;
+
+/// An error or inconsistency discovered when checking proof-carrying
+/// code.
+#[derive(Debug, Clone)]
+pub enum PccError {
+    /// An operation wraps around, invalidating the stated value
+    /// range.
+    Overflow,
+    /// An input to an operator that produces a fact-annotated value
+    /// does not have a fact describing it, and one is needed.
+    MissingFact,
+    /// A memory access is out of bounds.
+    OutOfBounds,
+    /// Proof-carry-code checking is not implemented for a certain case.
+    Unimplemented,
+}
+
+/// A fact on a value.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub enum Fact {
+    /// A bitslice of a value (up to a bitwidth) is less than or equal
+    /// to a given maximum value.
+    ///
+    /// The slicing behavior is needed because this fact can describe
+    /// both an SSA `Value`, whose entire value is well-defined, and a
+    /// `VReg` in VCode, whose bits beyond the type stored in that
+    /// register are don't-care (undefined).
+    ValueMax {
+        /// The bitwidth of bits we care about, from the LSB upward.
+        bit_width: u16,
+        /// The maximum value that the bitslice can take (inclusive).
+        max: u64,
```

Comment:
It is sort of already defined by the fact we are using a `u64` for `max`, but it might be good to reiterate that we are interpreting values as unsigned here.

Reply:
Done!